Sources/Afluent/Workers/Retry.swift

```diff
@@ -107,6 +58,74 @@ extension AsynchronousUnitOfWork {
     ///
     /// - Returns: An `AsynchronousUnitOfWork` that emits the same output as the upstream but retries on the specified error up to the specified number of times.
     public func retry<E: Error & Equatable>(_ retries: UInt = 1, on error: E) -> some AsynchronousUnitOfWork<Success> {
-        Workers.RetryOn(upstream: self, retries: retries, error: error)
+        Workers.Retry(upstream: self, strategy: RetryByCountOnErrorStrategy(retryCount: retries, error: error))
+    }
+}
+
+public protocol RetryStrategy: Sendable {
+    func handle(error: Error, beforeRetry: @Sendable (Error) async throws -> Void) async throws -> Bool
+}
+
+extension RetryStrategy {
+    func handle(error err: Error) async throws -> Bool {
+        try await handle(error: err, beforeRetry: { _ in })
+    }
+}
+
+extension RetryStrategy where Self == RetryByCountStrategy {
+    public static func byCount(_ count: UInt) -> RetryByCountStrategy {
+        return RetryByCountStrategy(retryCount: count)
+    }
+}
+
+public actor RetryByCountStrategy: RetryStrategy {
+    var retryCount: UInt
+    
+    public init(retryCount: UInt) {
+        self.retryCount = retryCount
+    }
+
+    public func handle(error err: Error, beforeRetry: @Sendable (Error) async throws -> Void) async throws -> Bool {
+        guard retryCount > 0 else {
+            return false
+        }
+        
+        try await beforeRetry(err)
+        decrementRetry()
+        return true
+    }
+    
+    func decrementRetry() {
+        guard retryCount > 0 else { return }
+        retryCount -= 1
+    }
+}
+
+public actor RetryByCountOnErrorStrategy<Failure: Error & Equatable>: RetryStrategy {
+    var retryCount: UInt
+    public let error: Failure
+    
+    public init(retryCount: UInt, error: Failure) {
+        self.retryCount = retryCount
+        self.error = error
+    }
+    
+    public func handle(error err: Error, beforeRetry: @Sendable (Error) async throws -> Void) async throws -> Bool {
+        guard retryCount > 0 else {
+            return false
+        }
+
+        guard !(err is CancellationError) else { throw err }
```

Comment:
What is the handled error is `CancellationError`? Should we support that case (does it make sense at all)?

Reply:
For right now? No, because the current strategy of Afluent across the library is to just handle cancellation for you. 

That said, I foresee a big change before v1 comes out that challenges that fundamental aspect of Afluent and creates operators that don't necessarily have a possible failure and forces the same kind of cancellation considerations as other async/await paradigms in Swift.